Gawron/ZPI-14-ZPI-58-BE-17-Dodanie-i-usunięcie-opiekuna-z-obserwowanych

Gawron97 opened the pull request.

Gawron97 committed:
add structure for following caretakers
add data generation for following caretakers

Gawron97 committed:
add logic for adding and removing caretakers from following
add tests

Gawron97 committed:
change ddl-auto to update

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/ClientController.java:
```diff
@@ -26,10 +25,35 @@ public class ClientController {
         return clientService.getClient(principal.getName());
     }
 
+    @Operation(summary = "Add caretaker to following list of client")
+    @PostMapping("/add-following-caretakers")
+    @PreAuthorize("isAuthenticated()")
+    public ClientComplexInfoDTO addFollowingCaretakers(Principal principal,
+                                                       @RequestParam Set<String> caretakerEmails,
+
+                                                       @RoleParameter
+                                                      @AcceptRole(acceptRole = Role.CLIENT)
+                                                      @RequestHeader(value = "${header-name.role}") Role role) {
+        return clientService.addFollowingCaretakers(principal.getName(), caretakerEmails);
+    }
```
Nie bardzo widzę jak frontend miałby użyć dodania paru osób na raz do obserwowania. Lepiej byłoby zrobić:

```java
@PostMapping("/follow/{caretakerEmail}")
``` 

Plus znacznie się url uprości

Gawron97 replied:
Rzeczywiście, ja sobie wyobrazilem ze normalnie moglby sobie z listy wybierac i kilka na raz dac, ale w sumie to nie bedzie listy tylko np przy profilu opiekuna dodanie

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/ClientController.java:
```diff
@@ -26,10 +25,35 @@ public class ClientController {
+        return clientService.addFollowingCaretakers(principal.getName(), caretakerEmails);
+    }
+
+    @Operation
+    @PostMapping("/remove-following-caretakers")
+    @PreAuthorize("isAuthenticated()")
+    public ClientComplexInfoDTO removeFollowingCaretakers(Principal principal,
+                                                          @RequestParam Set<String> caretakerEmails,
+
+                                                          @RoleParameter
+                                                          @AcceptRole(acceptRole = Role.CLIENT)
+                                                          @RequestHeader(value = "${header-name.role}") Role role) {
+        return clientService.removeFollowingCaretakers(principal.getName(), caretakerEmails);
+    }
```
Analogicznie tutaj zrobiłbym 

```java
@DeleteMapping("/follow/{caretakerEmail}")
```

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/ClientController.java:
Ogarnąłbym wcięcia w argumentach funkcji żeby były podobne wszędzie

lukcyn commented on src/main/java/com/example/petbuddybackend/dto/user/ClientComplexInfoDTO.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Set;
+
+@Builder
+public record ClientComplexInfoDTO(
+        AccountDataDTO accountData,
+        Set<String> followingCaretakersEmails
+) {
+}
```
> AccountDataDTO accountData

Zastanowiłbym się czy to nie przenieść do ClientDTO (`GET /api/client`) a w endpointach z followami tylko ten `Set<String> zwracać`. Szcególnie, że `AccountDataDTO` ma np. zdjęcie profilowe które trzeba odnawiać i imo to jest niepotrzebny overhead

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/user/Client.java:
```diff
@@ -25,4 +28,25 @@ public class Client {
+    @Builder.Default
+    private Set<Caretaker> followingCaretakers = new HashSet<>();
+
+    @PreUpdate
+    public void preUpdate() {
+        assertClientNotFollowingItself();
+    }
+
+    private void assertClientNotFollowingItself() {
+        if (followingCaretakers.stream().anyMatch(caretaker -> caretaker.getEmail().equals(email))) {
+            throw new IllegalActionException("Client cannot follow itself");
+        }
+    }
```
Bez PrePersist ci może nie ztriggerować tego asserta

Plus to jest dobry kandydat na `@Check` więc można (a nawet myślę że warto by było) to na poziomie bazy danych sprawdzać

Gawron97 replied:
moge w Check sprawdzic cala liste?

lukcyn commented on src/main/java/com/example/petbuddybackend/service/mapper/ClientMapper.java:
```diff
@@ -1,15 +1,33 @@
 
     ClientDTO mapToClientDTO(Client client);
 
+    @Mapping(target = "followingCaretakersEmails", source = "client.followingCaretakers", qualifiedByName = "mapFollowingCaretakersEmails")
+    ClientComplexInfoDTO mapToClientComplexInfoDTO(Client client);
+
+    @Named("mapFollowingCaretakersEmails")
+    default Set<String> mapFollowingCaretakersEmails(Set<Caretaker> followingCaretakers) {
+        return followingCaretakers
+                .stream()
+                .map(caretaker -> caretaker.getAccountData().getEmail())
+                .collect(Collectors.toSet());
+    }
```
Dodałbym test czy nie ma pól jako null. Podobne jak w innych testach z mapperami

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/ClientService.java:
```diff
@@ -47,6 +55,34 @@ public void createClientIfNotExist(JwtAuthenticationToken token) {
+        return clientMapper.mapToClientDTO(client);
+    }
+
+    public ClientComplexInfoDTO addFollowingCaretakers(String clientEmail, Set<String> caretakerEmails) {
```
Tak jak w kontrolerze. Myślę, że jakby dodać zamiast Set<String> po prostu String caretakerEmail to sie uprości

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/ClientService.java:
```diff
@@ -47,6 +55,34 @@ public void createClientIfNotExist(JwtAuthenticationToken token) {
+        return clientMapper.mapToClientComplexInfoDTO(clientRepository.save(client));
+    }
+
+    public ClientComplexInfoDTO removeFollowingCaretakers(String clientEmail, Set<String> caretakerEmails) {
```
> Tak jak w kontrolerze. Myślę, że jakby dodać zamiast Set po prostu String caretakerEmail to sie uprości

Jak wyżej

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/ClientService.java:
```diff
@@ -47,6 +55,34 @@ public void createClientIfNotExist(JwtAuthenticationToken token) {
+
+        Client client = getClientByEmail(clientEmail);
+        assertClientNotFollowingItself(client, caretakerEmails);
+        assertNewFollowingCaretakersAreNotAlreadyFollowed(client, caretakerEmails);
```
Pytanie czy trzeba to sprawdzać. Myślę, że jak trafi się duplikat i chce mu znowu dać follow, to ostatecznie osoby które miały być zaobserwowane są zaobserwowane

Gawron97 replied:
Niby tak, ale zwrocilbym informacje, ze opiekun jest juz zaobserwowany

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/ClientService.java:
```diff
@@ -47,6 +55,34 @@ public void createClientIfNotExist(JwtAuthenticationToken token) {
+
+        Client client = getClientByEmail(clientEmail);
+        assertClientNotFollowingItself(client, caretakerEmails);
+        assertCaretakersToRemoveAreFollowed(client, caretakerEmails);
```
Tu podobnie. Jak chcemy usunąć kogoś kogo nie ma to wydaje mi się, że API może zwrócić OK

Gawron97 replied:
Tutaj tez bym zwrocil info jesli unfollowujemy kogos kogo nie followujemy, chociaz i tak front nie dopusci pewnie do tego

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/ClientController.java:
```diff
@@ -26,10 +25,35 @@ public class ClientController {
+        return clientService.addFollowingCaretakers(principal.getName(), caretakerEmails);
+    }
+
+    @Operation
```
Tutaj nie powinien być opis jakiś?

lukcyn reviewed (commented):
Sugestie jak można to zrobić troche prościej. Jeszcze w komentarzu pisałem o zwracaniu tego `Set<String> followingCaretakersEmails` ale może w ogóle go nie trzeba zwracać tylko proste OK. Być może dzięki temu nie będzie trzeba robić GET na opiekunów za każdym obserwowaniem

Gawron97 committed:
returning Set of actual followed caretaker from add and delete endpoints
refactor to add and delete only one caretaker at the one time

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/user/Client.java:
```diff
@@ -25,4 +28,25 @@ public class Client {
+    @Builder.Default
+    private Set<Caretaker> followingCaretakers = new HashSet<>();
+
+    @PreUpdate
+    public void preUpdate() {
+        assertClientNotFollowingItself();
+    }
```
```suggestion
    @PrePersist
    @PreUpdate
    public void preSave() {
        assertClientNotFollowingItself();
    }
```

Gawron97 replied:
tak jest tak jest

lukcyn commented on src/main/java/com/example/petbuddybackend/entity/user/Client.java:
```diff
@@ -25,4 +28,25 @@ public class Client {
     private List<Rating> ratings = new ArrayList<>();
+
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "following_caretakers",
+            joinColumns = @JoinColumn(name = "client_email"),
+            inverseJoinColumns = @JoinColumn(name = "caretaker_email")
+    )
```
```suggestion
    @JoinTable(
            name = "following_caretakers",
            joinColumns = @JoinColumn(name = "client_email", nullable = false),
            inverseJoinColumns = @JoinColumn(name = "caretaker_email", nullable = false)
    )
```

lukcyn reviewed (approved):
Dwa komentarze, a poza tym jest git

Gawron97 committed:
add prePersist
make client_email and caretaker_email in following_caretakers table not null
add unique constraint to following_caretaker table
